File: python/0230.py
from typing import Optional
from data_strcutres import TreeNode


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        return self.morris(root, k)
        #  return self.iterative(root, k)
        #  return self.recursive(root, k)

    def morris(self, root: Optional[TreeNode], k: int) -> int:
        lst = []
        while root:
            if root.left:
                prev = root.left
                while prev.right and prev.right != root:
                    prev = prev.right
                if not prev.right:
                    prev.right = root
                    root = root.left
                else:
                    prev.right = None
                    lst.append(root.val)
                    if len(lst) >= k:
                        break
                    root = root.right
            else:
                lst.append(root.val)
                if len(lst) >= k:
                    break
                root = root.right
        return lst[k-1]


    def iterative(self, root: Optional[TreeNode], k: int) -> int:
        stack, lst = [], []
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            lst.append(root.val)
            if len(lst) >= k:
                break
            root = root.right
        return lst[k-1]

    def recursive(self, root: Optional[TreeNode], k: int) -> int:
        def helper(node, lst):
            if node:
                helper(node.left, lst)
                lst.append(node.val)
                if len(lst) >= k: return
                helper(node.right, lst)
        lst = []
        helper(root, lst)
        return lst[k-1]

root = TreeNode(3)
root.left = TreeNode(1)
root.left.right = TreeNode(2)
root.right = TreeNode(4)

print(Solution().kthSmallest(root, 1))

File: cpp/0230.cpp
#include <vector>
#include <queue>
#include <iostream>
#include <functional>

using namespace std;



struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v) {}
    vector<int> treeToVec() {
        vector<int> vec{};
        queue<TreeNode*> que;
        que.push(this);
        while (!que.empty()) {
            TreeNode* node = que.front();
            que.pop();
            if (node) {
                vec.push_back(node->val);
                que.push(node->left);
                que.push(node->right);
            } else {
                vec.push_back(-1);
            }
        }
        return vec;
    }
};



class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int res{};
        const static function<void(TreeNode*)> traverse = [&](TreeNode* node) {
            if (!node) return;
            traverse(node->left);
            k--;
            if (k == 0) {
                res = node->val;
                return;
            }
            traverse(node->right);
        };
        traverse(root);
        return res;
    }
};
